### ISSUE TITLE
Missing initialization for map variables.

### ISSUE THREAD

**Issue by michael-emmi** [SRI-CSL/solidity#45]
@beillahi and I noticed that the verifier does not assume default initial values for map variables:
```
$ cat test.sol && solc-verify.py test.sol 
pragma solidity ^0.5.0;

/// @notice invariant m[0] == 0
contract Test {
    mapping(int => int) m;
}
Test::[implicit_constructor]: ERROR
 - test.sol:4:1: State variable initializers might violate invariant 'm[0] == 0'.
Test::[receive_ether_selfdestruct]: OK
Errors were found by the verifier.
```
The assumption could be expressed in Boogie like this:
```
$ cat test.bpl && boogie test.bpl /useArrayTheory
function {:builtin "((as const (Array Int Int)) 0)"} I() returns ([int] int);

procedure test() {
    assert(I()[42] == 0);
}
Boogie program verifier version 2.4.1.10503, Copyright (c) 2003-2014, Microsoft.

Boogie program verifier finished with 1 verified, 0 errors
```
Seems like a no brainer :-D

**Comment by dddejan** [comment]
This relies on Z3 specific support for constant arrays.

**Comment by hajduakos** [comment]
Just tried it with CVC4 (version 1.6):
```
$ boogie Test.bpl /useArrayTheory /trace /proverOpt:SOLVER=CVC4 /cvc4exe:/usr/local/bin/cvc4
```
Seems to work:
```
Boogie program verifier version 2.4.1.10503, Copyright (c) 2003-2014, Microsoft.
Parsing Test.bpl
Coalescing blocks...
Inlining...

Running abstract interpretation...
  [0.022974 s]
[TRACE] Using prover: /usr/local/bin/cvc4

Verifying test ...
  [0.151 s, 1 proof obligation]  verified

Boogie program verifier finished with 1 verified, 0 errors
```

**Comment by dddejan** [comment]
I might be wrong, I thought they don't support constant arrays. Can you paste the smt2 output.

**Comment by hajduakos** [comment]
```
(set-option :print-success false)
(set-info :smt-lib-version 2.0)
(set-option :produce-models true)
(set-logic ALL_SUPPORTED)
; done setting options


(declare-fun ControlFlow (Int Int) Int)
(declare-fun %lbl%+0 () Bool)
(declare-fun %lbl%@1 () Bool)
(declare-fun %lbl%+2 () Bool)
(push 1)
(set-info :boogie-vc-id test)
(assert (not
 (=> (= (ControlFlow 0 0) 36) (let ((anon0_correct  (=> (and %lbl%+0 true) (or %lbl%@1  (=> (= (ControlFlow 0 29) (- 0 46)) (= (select ((as const (Array Int Int)) 0) 42) 0))))))
(let ((PreconditionGeneratedEntry_correct  (=> (and %lbl%+2 true) (=> (= (ControlFlow 0 36) 29) anon0_correct))))
PreconditionGeneratedEntry_correct)))
))
(check-sat)
(pop 1)
; Valid
```

**Comment by dddejan** [comment]
That would be great. Let me do some more checking tonight.

**Comment by dddejan** [comment]
https://github.com/CVC4/CVC4/issues/1667

**Comment by hajduakos** [comment]
Tried it for a bit more complicated example with nested maps and data types.
```
// Simple int array
function {:builtin "((as const (Array Int Int)) 0)"} def_int_int() returns ([int] int);

// Nested array (int, int)
function {:builtin "((as const (Array Int (Array Int Int))) ((as const (Array Int Int)) 0))"} def_int_int_int() returns ([int][int] int);

// Array of structs
// struct T { int z; }
type {:datatype} Tx;
function {:constructor} TC(z : int) returns (Tx);
function {:builtin "((as const (Array Int T@Tx)) (TC 0))"} def_int_T() returns ([int] Tx);

procedure test() {
    var x : [int]int;
    var y : [int]int;
    assert(def_int_int()[42] == 0);
    assert(def_int_int_int()[1][2] == 0);
    assert(z#TC(def_int_T()[0]) == 0);

    x := def_int_int();
    y := def_int_int();

    assert(x[0] == y[0]);
    assert(x == y);
}
```

**Comment by hajduakos** [comment]
Seems to be working for these hand-made examples, but the hacky part is that we have to assemble these SMT strings "by hand", bypassing the translation of Boogie. This seems a bit fragile for types. For example, Boogie translates a custom datatype `Tx` to `T@Tx`, which we also have to do manually in the SMT formulas.

**Comment by hajduakos** [comment]
Implemented in af587c0eb03b0614dbb9514f034ceab53533a891 and seems to be working for examples including nested arrays, mappings, structs and even for simple equalities connecting them. The Boogie specific parts are located in the new `smtname` member of `boogie::TypeDecl` and the `ASTBoogieUtils::defaultValueInternal` function. Although this parts are a bit fragile to changes in Boogie, I prefer this solution because we could remove our complicated approach of getting possible keys and initializing arrays/maps there (which does not work for the example in the issue).

**Comment by michael-emmi** [comment]
Works @beillahi and I!
